HIP/hip-991.md

```diff
@@ -44,182 +44,310 @@ This document outlines the development of a fixed fee system for the submission
+* **Topic Operator:** Entities or individuals who manage and operate topics, interested in reducing the noise of public topic IDs/making it more difficult to spam an open topic.
+* **Regular User:** End users who interact with topics, adding paid methods of interaction and submission to topics.
 
-### **FixedFee**
+**User Stories:**
 
-The `FixedFee` message represents a fixed fee charged for each message submission to a topic. This protobuf message already exists so it can be re-used.
+* As a **Topic Operator**, I want to set a fixed fee for submitting messages to my topic, so that I can ensure a consistent revenue stream and manage my topic effectively.
+* As a **Regular User**, I want to know the exact cost of interacting with a topic.
 
-```
-message FixedFee {
-  // The number of units to assess as a fee
-  int64 amount = 1;
-  // The denomination of the fee; taken as hbar if left unset
-  TokenID denominating_token_id = 2;
-}
-```
+## Specification
 
-### **CustomFixedFee**
+### Overview
 
-The `CustomFixedFee` message defines the type of fee and the account receiving the fee assessed during a message submission to the associated topic. A custom fee may only be fixed and must specify a fee collector account to receive the assessed fees.
+We propose adding a fixed fee mechanism to the Hedera Consensus Service (HCS) for topic messages, similar to the custom fee structures in the Hedera Token Service (HTS). This will involve creating new protocol buffers (protobuf) messages and modifying existing ones to accommodate fixed fees for topics.
 
-```
-message CustomFixedFee {
-  // Fixed fee to be charged
-  FixedFee fixed_fee = 1;
+### Requirements and Scope
 
-  // The account to receive the custom fee
-  AccountID fee_collector_account_id = 3;
-}
-```
+* Topics can have a Fee Schedule Key set at creation time.
+* Fee Schedule Key can manage fee updates and send free messages to the topic.
+* The Fee Schedule Key can be updated with an update topic transaction signed by the Admin Key and the new Fee Schedule Key.
+* If the topic was created without a Fee Schedule Key, the key cannot be added later.
+* Topics can have fees for submitting messages.
+* Fees can be set at creation time.
+* Fees can be changed (updated or removed) from a topic with an update topic transaction signed by the Fee Schedule Key.
+* Fees are defined as a list of custom fee.
+* A custom fee can be set in HBAR or HTS fungible tokens, and must have an accountID as a receiver.
+* Fee payer must set an allowance in HBAR or HTS tokens to be used to pay fees for message submissions, on an HCS topic basis.
+* Allowance granted to topic ID does not allow the topic's Admin Key (or any of the topic's keys) to spend that allowance.
+* Fee payer must set can set a maximum fee per message, on an HCS topic basis.
+* Messages signed by the Fee Schedule Key or by a Submit Key don't incur fees.
```

Comment:
Should the fee schedule key be signing transactions of messages? I think only the submit key, and if the account is the collector of the fee / or the operator of the topic should they be exempt.

Reply:
Yeah, that was my initial thought also, but it doesn't work well.

The point is that if a topic has a submit key, only the submit key can sign valid message transactions for that topic. Nobody else can send messages on the topic.

So, if you don't set a supply key, everyone can send messages to the topic, but now you no longer have a "free" messages key.

Because I assume the latter is the use case the HIP wants to address, I introduced the free messages for the Fee Schedule Key.

What do you think?

Reply:
Gotcha, that is a great point. For clarity, the operator who creates the topic and the fee collectors  on the custom fee will NOT be charged, so you would still have a free submission from the creator of the topic. 

But you would not have an ability to adhoc sign transactions from other accounts you may also want to remove fees from. I think there's enough of a use case for the need of the custom fee key to be used 

I agree, allowing the fee schedule key to sign to not charge fees is a good feature

Reply:
Good! 👍

Another topic: the HIP does not currently include the creators and fee collectors in the list of signers for free messages. We need to specify this clearly if we want it to.

**Creator**

Technically, we don't have that concept of the token's creator, but a topic can have an admin key (if it's set at creation time), so maybe we can add the admin key to the list of valid signers for free messages.

What do you think?

**Fee collectors**

Regarding fee collectors, I suggest reconsidering having them as free message signers for some reason, but in particular because I assume fee collectors must remain free not to trust each other.

Scenario
- Fees are set as 1 TokenA to Alice and 2 TokenB to Bob.
- Bob and Alice do not necessarily like or trust each other. Or they do now, but not in the future.
- Bob still wants his part of the fees when Alice sends messages.
- Alice can still technically send free messages for her part (because her balance in TokenA will not change), but she still needs to pay Bob's cut of the pie with TokenB.
- The scenario seems reasonable and predictable.

If we introduce free messages for any fee collectors in general, Alice can fool Bob, for example, by trading his fees OTC to her users: Bob will never see payments collected.

Even in a non-fraud scenario, Alice might be able to afford to send X number of messages without collecting the fee (i.e., she wants to promote her application), but this may not be the case for Bob.

I think fee receivers are a different concept from free message senders, so if we still feel there's a business need for a free message key list, I suggest keeping them separated.

We can introduce one explicit list, which only the admin key can update. 

In this way, as an admin, you gain the ability to make messages free for some users, and you make the rules check simpler and clearer.

Because the list can be updated, the free messages for specific users can be set for a limited time, so it would be more flexible instead of being assigned by your role.

Finally, this approach keeps the trust assumptions the same because Alice and Bob need to constantly trust the admin in any case to set their fee correctly.

What do you think?